Contract address page

technophile-04 opened the pull request:
### Description : 

Maybe let's keep the goal of this PR short, displaying the user entered contract details (verified + custom) using SE-2 
ContractUI component properly so that we can do read and write operations. 

**DONE :** 
- [x] Fetching contract details for verfied contract and displaying using SE-2 contract UI (Port, Shiv)
- [x] Making it multichain i.e making `network` from query param as targetNetwork (currently SE-2 is kind of configured to only look for scaffold.config targetNetworks )  (@technophile-04)
- [x] Header (Pablo)
- [x] Allowing uses to paste in Custom address and abi and show those details (Maybe we can use zustand global store for handling this) (port)
- [x] Handle loading, error states properly and display them (@Pabl0cks)

technophile-04 commented:
Update: completed first todo from https://github.com/BuidlGuidl/abi.ninja/pull/31#issue-2054192963 


https://github.com/BuidlGuidl/abi.ninja/assets/80153681/bb4c1424-8345-4b6f-a494-0c6fc8815cd5


Also updated https://github.com/BuidlGuidl/abi.ninja/pull/31#issue-2054192963 with some todo's, please feel free to edit that comment, move it to WIP if anyone is tackling one of it 🙌

technophile-04 commented:
Update completed 2nd todo at [616fdfe](https://github.com/BuidlGuidl/abi.ninja/pull/31/commits/616fdfefc222deda9a0e939347d1fcf4160c3f99) & [f531d0b](https://github.com/BuidlGuidl/abi.ninja/pull/31/commits/f531d0b069bf6a41c6420c0a51b1883a4c3f285f) : 


https://github.com/BuidlGuidl/abi.ninja/assets/80153681/b8c43d8f-bdcc-47f7-b11a-65d2509ae241


Currently showing sepolia, goerli and miainnet (since we had its api key already in SE-2) to add new networks we just need add extra chain to `targetNetworks` array in `scaffold.config.ts` and add `apiKey` and `etherscanEndpoint` in `NETWORKS_EXTRA_DATA`  for that chain id inside `utils/scaffold-eth/networks.ts` and it should start working 🙌

Pabl0cks commented on packages/nextjs/utils/scaffold-eth/networks.ts:
```diff
@@ -37,9 +152,13 @@ export const NETWORKS_EXTRA_DATA: Record<string, ChainAttributes> = {
   },
   [chains.optimism.id]: {
     color: "#f01a37",
+    etherscanEndpoint: "https://api-goerli-optimism.etherscan.io",
+    etherscanApiKey: MAIN_ETHERSCAN_API_KEY,
```
Probably a copy&paste typo here, I think it should be: `https://api-optimistic.etherscan.io`

I'm also having problems using `MAIN_ETHERSCAN_API_KEY` for OP Mainnet, I had to create an Optimistic Etherscan account & API to make it work. 

Should we create a `OPTIMISTIC_ETHERSCAN_API_KEY` with a shared API for development?

technophile-04 replied:
Ohh yes, tysm for the catch !! updated them also added some new networks 

also updated `.env.example`,  I can share my `.env.local` on TG DM for testing, also updated the main comment with TODO to add those in vercel environment 🙌

Pabl0cks commented on packages/nextjs/utils/scaffold-eth/networks.ts:
```diff
@@ -37,9 +152,13 @@ export const NETWORKS_EXTRA_DATA: Record<string, ChainAttributes> = {
   },
   [chains.optimismGoerli.id]: {
     color: "#f01a37",
+    etherscanEndpoint: "https://api-goerli-optimism.etherscan.io",
+    etherscanApiKey: MAIN_ETHERSCAN_API_KEY,
```
Endpoint and ApiKey are working fine (which is strange, because Optimistic API is working also for this endpoint).

In [docs](https://docs.optimism.etherscan.io/getting-started/endpoint-urls) they referencing `https://api-goerli-optimistic.etherscan.io` as endpoint, should we change it and also use Optimistic Etherscan API to make it more consistant?

Pabl0cks reviewed (commented):
Amazing job @technophile-04 with the last features, still testing them, but they are looking really good 👌

Left a tiny review because I was receiving errors while trying to load Optimism Mainnet contracts, will continue testing 🙌

technophile-04 commented on packages/nextjs/pages/[contractAddress]/[network].tsx:
```diff
@@ -1,18 +1,99 @@
+      <MiniHeader />
+      <div className="flex flex-col gap-y-6 lg:gap-y-8 justify-center items-center">
+        {isLoading ? (
+          <Spinner />
```
@Pabl0cks should we use : 

```tsx
<span className="loading loading-spinner text-primary h-14 w-14"></span>
```

Or maybe `<Spinner height="100" width="100"  />` so that it looks a bit big ? 

Also we updated <Spinner /> component internal colors 

lol sorry for nitpick we can always improve this later

Pabl0cks replied:
I think not using Spinner is more consistent with `index.tsx` behavior, thanks for the feedback ♥

Agree it was also too small :) Just pushed your suggestion 🙌

technophile-04 commented on packages/nextjs/pages/index.tsx:
```diff
@@ -3,51 +3,108 @@ import Image from "next/image";
+import { fetchContractABIFromEtherscan, getNetworksWithEtherscanApi, parseAndCorrectJSON } from "~~/utils/abi";
+import { notification } from "~~/utils/scaffold-eth";
+
+export const publicClient = createPublicClient({
+  chain: chains.mainnet,
+  transport: http(),
+});
```
@portdeveloper maybe we should use `usePublicClient({chainId : parseInt(network)})` instead because we want to check on selected chain if its contract or not currently we are always checking on mainnet irrespective of network selected

portdeveloper replied:
Oh thanks 🫡

technophile-04 commented on packages/nextjs/pages/index.tsx:
```diff
@@ -3,51 +3,108 @@ import Image from "next/image";
 
+const networks = getNetworksWithEtherscanApi();
+
+const isContractAddress = async (address: Address) => {
```
maybe we could move it to `utils/abi.ts`

technophile-04 commented:
Just added a couple of minor comments but I think we are almost there !! Tysm guys, really awesome work !!

---

Just added a small utility function to parse abi at [83f1d73](https://github.com/BuidlGuidl/abi.ninja/pull/31/commits/83f1d73d237e44215bbfda69e1197f4e1e5d7a60) (thanks to chatGPT) if it was pasted with non json stringified / normal abi array 

Test address : `0xd7036EDDDA3743Aa577DC24ca817f3F85099bDbD`

<details>
<summary>
Test Abi prettier formated : 
</summary>

```ts
[
  {
    inputs: [
      {
        internalType: "address",
        name: "_owner",
        type: "address",
      },
    ],
    stateMutability: "nonpayable",
    type: "constructor",
  },
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "previousOwner",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "newOwner",
        type: "address",
      },
    ],
    name: "OwnershipTransferred",
    type: "event",
  },
  {
    inputs: [],
    name: "increaseCounter",
    outputs: [],
    stateMutability: "payable",
    type: "function",
  },
  {
    inputs: [],
    name: "owner",
    outputs: [
      {
        internalType: "address",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "renounceOwnership",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [],
    name: "totalCounter",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "newOwner",
        type: "address",
      },
    ],
    name: "transferOwnership",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "",
        type: "address",
      },
    ],
    name: "userGreetingCounter",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "withdraw",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    stateMutability: "payable",
    type: "receive",
  },
]
```
</details>

Select network `sepolia` and paste the above things in custom Address + abi tab to test it. Would love if someone could give it some basic testing 🙌

PS: In actual abi.ninja if you paste the above abi it should give you an error, but I think it's a good feature to allow people to paste in direct abi array

technophile-04 reviewed (commented):
Look solid 💪 ! Tysm All !!! Just did a final pass and it works great !! Would love if come could give it a test too, We can merge it tomorrow 🙌

Pabl0cks reviewed (approved):
I think it's working nice and it's good to merge and continue iterating. Great job everyone!! 🙌

I'm trying to reproduce some small issues with error handling, but I'll add an issue to tackle it in the next PR.

Pabl0cks commented:
I was having a problem when refreshing a ContractDetailPage, it was always loading the "Please enter a valid address" on the first load, and then refreshing to the correct contract data loaded, or to another error in the case there was an actual error. I think `router.query` may be leaving `contractAddress` and `network` as `undefined` in the first render.

It didn't happen if I came directly from index, only when I enter the URL manually or refresh a loaded URL.

Just pushed a fix on bff17e3bb0aa204f26004631db7f2eeeb064cc41 but would love if another pair of eyes check it won't affect any working behavior.

technophile-04 commented:
Ohh yeah !! there was a quick flicker that showed whenever you refreshed the page, thanks for the fix lgtm !!

Tysm all merging this 🚀!!
